posthog/async_migrations/migrations/0005_person_collapsed_by_version.py

```diff
@@ -0,0 +1,252 @@
+                rollback=None,
+            ),
+            AsyncMigrationOperationSQL(
+                database=AnalyticsDBMS.CLICKHOUSE,
+                sql=f"""
+                    RENAME TABLE
+                        {PERSON_TABLE_NAME} to {BACKUP_TABLE_NAME},
+                        {TEMPORARY_TABLE_NAME} to {PERSON_TABLE_NAME}
+                    ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'
+                """,
+                rollback=f"""
+                    RENAME TABLE
+                        {PERSON_TABLE_NAME} to {FAILED_PERSON_TABLE_NAME},
+                        {BACKUP_TABLE_NAME} to {PERSON_TABLE_NAME}
+                    ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'
+                """,
+            ),
+            AsyncMigrationOperationSQL(database=AnalyticsDBMS.CLICKHOUSE, sql=PERSONS_TABLE_MV_SQL, rollback=None),
+            AsyncMigrationOperation(fn=optimize_table_fn),
+        ]
+
+    def new_table_engine(self):
+        engine = ReplacingMergeTree("person", ver="version")
+        # :TRICKY: Zookeeper paths need to be unique each time we run this migration, so generate a unique prefix.
+        engine.set_zookeeper_path_key(now().strftime("am0005_%Y%m%d%H%M%S"))
+        return engine
+
+    @cached_property
+    def pg_copy_target_person_id(self) -> int:
+        # :TRICKY: We calculate the last ID to copy at the start of migration once and cache it.
+        #    If the migration gets e.g. restarted, this is recalculated.
+        try:
+            return Person.objects.latest("id").id
+        except Person.DoesNotExist:
+            return -1
+
+    def get_pg_copy_highwatermark(self) -> int:
+        highwatermark = get_client().get(REDIS_HIGHWATERMARK_KEY)
+        return int(highwatermark) if highwatermark is not None else 0
+
+    def unset_highwatermark(self) -> None:
+        get_client().delete(REDIS_HIGHWATERMARK_KEY)
+
+    def copy_persons_from_postgres(self, query_id: str):
+        should_continue = True
+        while should_continue:
+            should_continue = self._copy_batch_from_postgres(query_id)
+
+    def _copy_batch_from_postgres(self, query_id: str) -> bool:
+        highwatermark = self.get_pg_copy_highwatermark()
+        if highwatermark > self.pg_copy_target_person_id:
+            logger.info(
+                "Finished copying people from postgres to clickhouse",
+                highwatermark=highwatermark,
+                pg_copy_target_person_id=self.pg_copy_target_person_id,
+            )
+            return False
+
+        persons = list(Person.objects.filter(id__gte=highwatermark)[:PG_COPY_BATCH_SIZE])
+        sql, params = self._persons_insert_query(persons)
+
+        execute_op_clickhouse(sql, params, query_id=query_id)
+
+        new_highwatermark = (persons[-1].id if len(persons) > 0 else self.pg_copy_target_person_id) + 1
+        get_client().set(REDIS_HIGHWATERMARK_KEY, new_highwatermark)
```

Comment:
nit: we have methods for get and unset but not set

Reply:
What do you mean? Redis client does have a `.set` function.

Reply:
I meant you wrote wrappers for when we get and unset but not for setting, that's inlined here